### PR TITLE
Added sign up functionality #9

### LINKED ISSUE #9: Allow users to signup
Verify email before allowing to login. By default logged in user cannot do anything extra.

### PR DESCRIPTION
**Pull request by HenriTammo**
Added Sign up buttons to the header and log in page.
Created a sign up form where you can enter your first and last name, email and password.
At the end of the form there is a checkbox that is required and a link to the terms of service.

closes #9

### CODE REVIEW

File: teemaderegister-react/src/components/HeaderWrap.jsx [hunk 1]
```diff
@@ -185,6 +185,9 @@ class HeaderWrap extends Component {
               <div className='login'>
                 <Link to='/login'>Sign in</Link>
               </div>}
+            <div className='login'>
+              <Link to='/sign_up'>Sign up &nbsp;</Link>
+            </div>
```
Comment (romilrobtsenkov) [inline teemaderegister-react/src/components/HeaderWrap.jsx:190]: I would rather keep the signup button only on login page, e.g something similar to this and remove from page header. 
Screenshot 2021-01-13 at 23 26 52

File: teemaderegister-react/src/components/SignUp.jsx [hunk 2]
```diff
@@ -0,0 +1,124 @@
...
+
+const { func, object } = PropTypes
+
+const propTypes = {
+  initSignUp: func.isRequired,
+  location: object.isRequired,
+  triggerSignUp: func.isRequired
+}
+
+class SignUp extends React.Component {
```
Comment (romilrobtsenkov) [inline teemaderegister-react/src/components/SignUp.jsx:21]: would prefer You would have used functional component, but this is ok

**Review by romilrobtsenkov — COMMENTED** [review]
There is no feedback after signup, form should be cleared and user should be notified that user creation was successful and he can log in

User gets no feedback if user email already exists

As both frontend and beckend change please update password complexity to require atleast one uppercase letter and a number

File: teemaderegister-react/src/actions/SignUpActions.js [hunk 3]
```diff
@@ -0,0 +1,20 @@
...
+import { SIGNUP_URL } from '../constants/ApiConstants'
+
+export const initSignUp = () => dispatch => dispatch({ type: types.SIGNUP_INIT })
+
+export const triggerSignUp = creds => dispatch => {
+  dispatch({ type: types.SIGNUP_START })
+
+  return Api('POST', SIGNUP_URL, { data: creds })
+    .then(data => {
+      setToken(data.token)
```
Comment (romilrobtsenkov) [inline teemaderegister-react/src/actions/SignUpActions.js:14]: no token is recieved, unneeded line

File: teemaderegister-react/src/components/SignUp.jsx [hunk 4]
```diff
@@ -0,0 +1,129 @@
...
+                <Input
+                  prefix={<UserOutlined />}
+                  type='name'
+                  placeholder='Last name'
+                />
+              </FormItem>
+              <FormItem name='email' rules={[
+                { required: true, message: 'Please input your email!' }
+              ]}>
+                <Input autoFocus prefix={<MailOutlined />} placeholder='Email' />
```
Comment (romilrobtsenkov) [inline teemaderegister-react/src/components/SignUp.jsx:94]: would make sense to move autofocus to first field in the form, e.g. first name

File: teemaderegister-react/src/components/SignUp.jsx [hunk 5]
```diff
@@ -0,0 +1,129 @@
...
+              </FormItem>
+              <FormItem name='password' className='signup__password' rules={[
+                { required: true, message: 'Please input your password!' },
+                { min: 8, message: 'Passwords must be at least 8 characters long' }
```
Comment (romilrobtsenkov) [inline teemaderegister-react/src/components/SignUp.jsx:98]: please add new password requirements here also

**Review by romilrobtsenkov — COMMENTED** [review]
For some reason previous message is always included. Try for example submitting same email for second time and then with new email. Both success message and error message are shown.

Screenshot 2021-01-17 at 19 35 34

File: teemaderegister-react/src/actions/SignUpActions.js (hunk 3 above)
Comment (SandraMai) [reply teemaderegister-react/src/actions/SignUpActions.js]: Removed

File: teemaderegister-react/src/components/SignUp.jsx (hunk 5 above)
Comment (SandraMai) [reply teemaderegister-react/src/components/SignUp.jsx:102]: ✓

File: teemaderegister-react/src/components/SignUp.jsx (hunk 4 above)
Comment (SandraMai) [reply teemaderegister-react/src/components/SignUp.jsx]: ✓

File: teemaderegister-react/src/components/HeaderWrap.jsx (hunk 1 above)
Comment (SandraMai) [reply teemaderegister-react/src/components/HeaderWrap.jsx]: ✓

**Review by romilrobtsenkov — APPROVED** [review]

**Pull request merged**